x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryTranslator.java

```diff
@@ -258,23 +250,35 @@ static String field(AggregateFunction af) {
             }
             return field.name();
         }
-        if (arg instanceof Literal) {
-            return String.valueOf(((Literal) arg).value());
-        }
         throw new SqlIllegalArgumentException("Does not know how to convert argument {} for function {}", arg.nodeString(),
                                               af.nodeString());
     }
 
-    private static String topAggsField(AggregateFunction af, Expression e) {
+    private static Object topAggsFieldOrScript(AggregateFunction af, Expression e) {
         if (e == null) {
             return null;
         }
         if (e instanceof FieldAttribute) {
             return ((FieldAttribute) e).exactAttribute().name();
         }
+        if (e instanceof ScalarFunction) {
+            return ((ScalarFunction) e).asScript();
+        }
         throw new SqlIllegalArgumentException("Does not know how to convert argument {} for function {}", e.nodeString(), af.nodeString());
     }
 
+    private static boolean isFieldOrLiteral(AggregateFunction af) {
+        return af.field().foldable() || af.field() instanceof FieldAttribute;
+    }
+
+    private static Object getFieldOrLiteralOrScript(AggregateFunction af) {
+        if (isFieldOrLiteral(af)) {
```

Comment:
`return  isFieldOrLiteral(af) ?  field(af) : ((ScalarFunction) af.field()).asScript()`

Reply:
The logic from `topAggsFieldOrScript` and this method is similar and should be reused - I like the instanceof check for Scalar instead of the direct cast; a bit verbose but safe.